### ISSUE TITLE
 关于TVA_fusion.py中的 t1, p, t2, n 

### ISSUE THREAD

**Issue by Jingluo0** [XpastaX/ConFEDE#16]
作者你好，我阅读了您的代码，关于TVA_fusion.py中的embedding pairs不太理解。
 if sample2 is not None:
                # For sequence [Ts, Vs, As, Td, Vd, Ad, T1s, V1s, ...... V6d, A6d], construct corresponding positive and negative embedding pairs, which is use for contrastive learning later.
                t1, p, t2, n = torch.tensor([0, 0, 7, 7, 14, 14,
                                             0, 0, 1, 1, 2, 2, 3, 3, 4, 4, 5, 5, 6, 6],
                                            device=device), \
                               torch.tensor([1, 2, 8, 9, 15, 16,
                                             7, 14, 8, 15, 9, 16, 10, 17, 11, 18, 12, 19, 13, 20],
                                            device=device), \
                               torch.tensor([0, 0, 0, 0, 7, 7, 7, 7, 14, 14, 14, 14,
                                             0, 0, 0, 1, 1, 1, 2, 2, 2, 3, 3, 3, 4, 4, 4, 5, 5, 5, 6, 6, 6],
                                            device=device), \
                               torch.tensor([3, 4, 5, 6, 10, 11, 12, 13, 17, 18, 19, 20,
                                             21, 28, 35, 22, 29, 36, 23, 30, 37, 24, 31, 38, 25, 32, 39, 26, 33, 40, 27,
                                             34, 41], device=device)

                indices_tuple = (t1, p, t2, n)
                pre_sample_label = torch.tensor([0, 0, 0, 1, 2, 3, 4, 0, 0, 0, 1, 2, 3, 4, 0, 0, 0, 1, 2, 3, 4,
                                                 5, 5, 5, 6, 7, 8, 9, 5, 5, 5, 6, 7, 8, 9, 5, 5, 5, 6, 7, 8, 9, ])

这里的 t1, p, t2, n， pre_sample_label 为什么要定义成这样，希望作者解惑。

**Comment by XpastaX** [comment]
这些值用来构建出对比学习的正负样本对。是hardcoding进代码的。顺序对应text， vision， audio的正负样本对。这样定义是为了适应pytorch-metric-learning==0.9.99 中计算对比学习loss的输入。

样本排列：
Ts, Vs, As, Td, Vd, Ad, T1s, V1s, ...... 

(t1[0], p[0]) 对应 正样本对 （Ts, Vs), 这两个的embedding会在学习中互相靠近

(t2[0], n[0]) 对应 副样本对 (Ts, Td), 这两个的embedding 会在学习中互相远离

每个样本会抽样六个其他样本来构成正负样本对，代码这里没有做自动化处理。hardcoding成 “构建7个样本之间的正负对"

**Comment by Jingluo0** [comment]
感谢您的解答！
不过对于排列Ts, Vs, As, Td, Vd, Ad, T1s, V1s, ......，之后的一些样本对，比如（t1[3], p[3])对应的是（V1s,T1d），这在论文公式中似乎不是对应正样本对。感觉t1=（0，0，6，6，12，12，......），p=（1，2，7，8，13，14,......）会更符合我的理解？
另外能否再解释一下pre_sample_label 的意思。
不胜感激！

**Comment by XpastaX** [comment]
> 9,

感谢指正。我发现这边comment 搞错了（之前改过comment，改之前是正确的）
[[根据]](https://github.com/XpastaX/ConFEDE/blob/169d38a0e33b35d7334967a373bf622e9b5cec30/SIMS/model/net/constrastive/TVA_fusion.py#L172)

实际上的排列应该是 Ts,T1s,T2s...T6s, Vs,V1s.....,As,A1s,...

我[重写一下这部分的代码](https://github.com/XpastaX/ConFEDE/blob/5858bf71245f67978757a6102bd54bf1c1b2745f/SIMS/config.py#L14)。硬编码不利于理解与纠错。

Pre_sample_label 的意义是指明哪些embedding之间应该互相接近，同标签的会接近，标签不同的应该远离。这个是pytorch-metric-learning要求的输入。

**Comment by Jingluo0** [comment]
好的，感谢您的解答和工作！

**Comment by DavidSmith-FHTT** [comment]
作者您好，同样在读了您的代码后，关于TVA_fusion.py中，对比损失的计算我不太懂，希望您解疑。
                indices_tuple = (t1, p, t2, n)

                pre_sample_label = torch.tensor([0, 0, 0, 1, 2, 3, 4, 0, 0, 0, 1, 2, 3, 4, 0, 0, 0, 1, 2, 3, 4,
                                                 5, 5, 5, 6, 7, 8, 9, 5, 5, 5, 6, 7, 8, 9, 5, 5, 5, 6, 7, 8, 9, ])
                for i in range(len(x1_all)):
                    pre_sample_x = []
                    for fea1, fea2 in zip([x_t_simi1, x_v_simi1, x_a_simi1, x_t_dissimi1, x_v_dissimi1, x_a_dissimi1, ],
                                          [x_t_simi2, x_v_simi2, x_a_simi2, x_t_dissimi2, x_v_dissimi2,
                                           x_a_dissimi2, ]):
                    pre_sample_x.append(torch.cat((fea1[i].unsqueeze(0), fea2[6 * i:6 * (i + 1)]), dim=0))

                    sup_const_loss += self.ntxent_loss(torch.cat(pre_sample_x, dim=0), pre_sample_label,
                                                       indices_tuple=indices_tuple)

                sup_const_loss /= len(x1_all)
1.请问pre_sample_label里面的具体数字[0, 0, 0, 1, 2, 3, 4, 0, 0, 0, 1, 2, 3, 4, 0, 0, 0, 1, 2, 3, 4,5, 5, 5, 6, 7, 8, 9, 5, 5, 5, 6, 7, 8, 9, 5, 5, 5, 6, 7, 8, 9, ]是如何得到的？
2. pre_sample_x.append(torch.cat((fea1[i].unsqueeze(0), fea2[6 * i:6 * (i + 1)]), dim=0))为什么构建这样的嵌入向量？也就是(fea1[i].unsqueeze(0), fea2[6 * i:6 * (i + 1)])的具体含义是什么？
3. 能否解疑一下ntxent_loss(torch.cat(pre_sample_x, dim=0), pre_sample_label,indices_tuple=indices_tuple)为什么这么使用？
期待您的回复，谢谢

**Comment by XpastaX** [comment]
> Ts,T1s,T2s...T6s, Vs,V1s.....,As,A1s,...
1. 相同标签意味着他们之间的embedding应该相同，不同标签之间对应的embedding应该远离。
2. 这样是为了 构建embedding的排列：Ts,T1s,T2s...T6s, Vs,V1s.....,As,A1s,...Td,T1d........A6d.  即首先收集所有sample的Ts，然后收集Vs，。。一直到Ad。数量上为 6种embedding 7个sample共 6*7=42个embedding，与pre_sample_label长度对应
3. 这个是因为计算ntxent_loss的package（pytorch-metric-learning）中就是这样定义的使用方式，所以前面的操作都是为了符合官方定义的输入格式，所以显得很复杂。

**Comment by DavidSmith-FHTT** [comment]
非常感谢您的解答
